ext/semian/semian.c

```diff
@@ -167,13 +134,11 @@ static VALUE
 update_ticket_count(update_ticket_count_t *tc)
 {
   short delta;
-  struct timespec ts = { 0 };
-  ts.tv_sec = kInternalTimeout;
 
   if (get_max_tickets(tc->sem_id) != tc->tickets) {
     delta = tc->tickets - get_max_tickets(tc->sem_id);
 
-    if (perform_semop(tc->sem_id, kIndexTickets, delta, 0, &ts) == -1) {
+    if (perform_semop(tc->sem_id, kIndexTickets, delta, 0) == -1) {
```

Comment:
Is one of the issues preventing this from going forward the lack of a `semtimedop` shim with timeouts for OS X? I looked into this. We'd need to bring back the nogvl lock.

Initially, I found @byroot 's shim here: https://github.com/Shopify/semian/pull/7#issuecomment-59564579 , but I found after running it on a live OS X install that using signals like `SIGALRM` won't work in a multi-threaded environment - a process can only schedule one alarm at a time, and alarm handlers don't easily distinguish between different threads, which are both big problems.

The way around this is to use some form of sleep. I wrote one that works (it can of course be improved on):
```c
#include <unistd.h> // for usleep

// timeval_subtract is adapted from 
// https://opensource.apple.com/source/distcc/distcc-31.0.81/src/timeval.c
int
timeval_subtract(struct timeval *result,
                 const struct timeval *x,
                 const struct timeval *y)
{
  struct timeval _x = *x;
  struct timeval _y = *y;

  /* Perform the carry for the later subtraction by updating Y. */
  if (_x.tv_usec < _y.tv_usec) {
    int nsec = (_y.tv_usec - _x.tv_usec) / 1000000 + 1;
    _y.tv_usec -= 1000000 * nsec;
    _y.tv_sec += nsec;
  }
  if (_x.tv_usec - _y.tv_usec > 1000000) {
    int nsec = (_x.tv_usec - _y.tv_usec) / 1000000;
    _y.tv_usec += 1000000 * nsec;
    _y.tv_sec -= nsec;
  }

  /* Compute the time remaining to wait.
     `tv_usec' is certainly positive. */
  result->tv_sec = _x.tv_sec - _y.tv_sec;
  result->tv_usec = _x.tv_usec - _y.tv_usec;

  /* Return 1 if result is negative. */
  return _x.tv_sec < _y.tv_sec;
}

int 
__attribute__((weak)) // GCC provides this, can also use an #ifdef
semtimedop(int semid, struct sembuf *sops, size_t nsops, const struct timespec *timeout)
{
  struct sembuf _sops = *sops;
  _sops.sem_flg |= IPC_NOWAIT;
  int code = semop(semid, &_sops, nsops);

  struct timeval timeout_date, current_date;
  gettimeofday(&timeout_date, NULL);
  timeout_date.tv_usec += timeout->tv_nsec/1000;
  while (timeout_date.tv_usec > 1000000) {
    timeout_date.tv_usec -= 1000000;
    timeout_date.tv_sec += 1;
  }
  timeout_date.tv_sec += timeout->tv_sec;

  int timed_out = 0;
  while (!timed_out && -1 == code && EAGAIN == errno) {
    errno = 0;
    code = 0;
    usleep(100); // Can be changed
    code = semop(semid, &_sops, nsops);
    gettimeofday(&current_date, NULL); // Get new time
    timed_out = timeval_subtract(&current_date, &timeout_date, &current_date); // See if elapsed time > timeout
  }
  return code;
}
```